extensions/amp-share-tracking/0.1/amp-share-tracking.js

```diff
@@ -15,43 +15,118 @@
  */
 
 import {isExperimentOn} from '../../../src/experiments';
-import {getService} from '../../../src/service';
+import {xhrFor} from '../../../src/xhr';
+import {viewerFor} from '../../../src/viewer';
 import {dev} from '../../../src/log';
 
 /** @private @const {string} */
 const TAG = 'amp-share-tracking';
 
-class AmpShareTracking extends AMP.BaseElement {
+export class AmpShareTracking extends AMP.BaseElement {
+  /**
+    * @return {boolean}
+    * @private
+    */
+  isExperimentOn_() {
+    return isExperimentOn(this.win, TAG);
+  }
+
+  /** @override */
+  isLayoutSupported(unusedLayout) {
+    return true;
+  }
 
   /** @override */
   buildCallback() {
-    this.isExperimentOn_ = isExperimentOn(this.win, TAG);
-    if (!this.isExperimentOn_) {
+    if (!this.isExperimentOn_()) {
       dev.warn(TAG, `TAG ${TAG} disabled`);
       return;
     }
+
+    /** @private {string} */
+    this.vendorHref_ = this.element.getAttribute('data-href');
+    dev.fine(TAG, 'vendorHref_: ', this.vendorHref_);
+
+    /** @private {!Promise<!Object>} */
+    this.shareTrackingFragments_ = Promise.all([
+      this.getIncomingFragment_(),
+      this.getOutgoingFragment_(this.vendorHref_)]).then(results => {
+        return {
+          incomingFragment: results[0],
+          outgoingFragment: results[1],
+        }
+      });
+
   }
-}
 
-/**
- * ShareTrackingService processes the incoming and outcoming url fragment
- */
-export class ShareTrackingService {
-  constructor(window) {
-    this.win = window;
+  /**
+   * Get the incoming share-tracking fragment from the viewer
+   * @return {!Promise<string>}
+   * @private
+   */
+  getIncomingFragment_() {
+    return viewerFor(this.win).getFragment().then(fragment => {
+      if (!fragment || fragment.indexOf('.') != 0) {
+        return Promise.resolve();
```

Comment:
Could I make it `Promise.resolve()` and make the return type `{!Promise<string|undefined>}`?